csrc/multidevice/utils.cpp

```diff
@@ -13,22 +13,88 @@
 #include <multidevice/lower_communication.h>
 #include <multidevice/utils.h>
 #include <ops/all_ops.h>
+#include <root_domain_map.h>
 #include <scheduler/utils.h>
 
 #include <c10/util/irange.h>
 
 namespace nvfuser {
 namespace {
 
-std::vector<IterDomain*> getShardedIterDomains(TensorView* tv) {
-  std::vector<IterDomain*> sharded_ids;
+std::unordered_set<IterDomain*> getShardedIterDomains(TensorView* tv) {
+  std::unordered_set<IterDomain*> sharded_ids;
   std::copy_if(
       tv->getLeafDomain().begin(),
       tv->getLeafDomain().end(),
-      std::back_inserter(sharded_ids),
+      std::inserter(sharded_ids, sharded_ids.begin()),
       [](auto id) { return id->isDeviceDim(); });
   return sharded_ids;
 }
+
+// Returns whether a IterDomain in a TensorView is the outermost
+// allocated IterDomain in the TensorView.
+bool isOutermostAllocatedId(TensorView* tv, IterDomain* id) {
+  for (auto i : tv->getLeafDomain()) {
+    if (i == id) {
+      return true;
+    }
+    if (!i->isDeviceDim() && !i->isReduction()) {
+      return false;
+    }
+  }
+  NVF_ERROR(
+      false, "Id", id->toString(), " is not in TensorView ", tv->toString());
+  return false;
+}
+
+// For a resharding expression, either a set or reduce, returns root IDs
+// that change sharding.
+// (1) sharded root IterDomains that are added by the expression
+// i.e. sharded IterDomains that are present in the output, but not the input.
+// (2) sharded root IterDomains that are removed by the expression
+// i.e. sharded IterDomains that are present in the input, but not the output.
+// TODO: Analyze leaf domain for unsharded/sharded IDs and return their
+// parent root IDs.
+std::pair<std::vector<IterDomain*>, std::vector<IterDomain*>> getShardingChanges(
+    Expr* expr) {
+  NVF_ERROR(
+      expr->outputs().size() == 1,
+      "Resharding operations can only have one output");
+  NVF_ERROR(
+      expr->inputs().size() == 1,
+      "Resharding operations can have only one input");
+  auto output = expr->outputs().at(0)->as<TensorView>();
+  auto input = expr->inputs().at(0)->as<TensorView>();
+
+  std::vector<IterDomain*> shard_additions;
+  std::vector<IterDomain*> shard_deletions;
+  auto rootmap = PairwiseRootDomainMap(input, output).mapBroadcast(false);
+  const auto c2p_map = rootmap.mapConsumerToProducer();
+
+  for (IterDomain* out_root : output->getRootDomain()) {
+    IterDomain* in_root = c2p_map.at(out_root);
+    // Ignore sharded reductions on the output
+    // ex. DIDx(i0) -> r(i0) or DIDx(i0) -> r(DIDx(i0))
+    // since they don't affect allocation.
+    if (in_root->isDeviceDim() && !out_root->isParallelized() &&
+        !out_root->isReduction()) {
+      shard_deletions.push_back(in_root);
+    } else if (!in_root->isParallelized() && out_root->isDeviceDim()) {
```

Comment:
Does this mean this function assumes no TIDX/BIDX parallelization is yet done?

Reply:
Correct.
Could something like this be allowed? `TID(i0)` -> `DIDx(i0)`
We could update to `(!in_root->isDeviceDim() && out_root->isDeviceDim())`

Reply:
That should have no problem.